### ISSUE TITLE
USDT tracing does not set semaphore

### ISSUE THREAD

**Issue by anelson** [iovisor/bpftrace#612]
TL;DR: USDT probes which make use of a semaphore to detect when probing is enabled are currently broken with the latest changes in `master`.

# Environment

```
$ uname -a
Linux turing 5.0.11-arch1-1-ARCH #1 SMP PREEMPT Thu May 2 19:48:15 UTC 2019 x86_64 GNU/Linux
$ bpftrace --version
bpftrace v0.9-145-ge959
$ pacman -Q bcc
bcc 0.9.0-1
```

# Problem Description

Some USDT probes make use of an optional feature of USDT probing, called "semaphores", in which a `unsigned short` variable placed in a special `.probes` section of the ELF binary is incremented whenever a probe is attached, and decremented when it's detached.  These probes can then check the value of this variable and only execute probing code if it's non-zero.

I'm working on a [Rust crate to bring USDT probing to stable  Rust](https://github.com/anelson/tracers), and due to the way it's implemented each time a probe is hit the Rust code has to `call` into a C++ function, and in some cases do even more work.  So I've implemented this crate to use semaphores to avoid this overhead unless a probe is enabled.

The problem is that `bpftrace` has a bug whereby it increments and then immediately decrements this semaphore, so the code firing the probe never sees a non-zero value and thus never fires the probe.

# Steps to reproduce

I modified the `usdt_test` test program in the `bpftrace` source tree to illustrate this problem.  Here's the code I'm using:

```c
#define _SDT_HAS_SEMAPHORES 1


#define STAP_HAS_SEMAPHORES 1 /* deprecated */

#include <sys/sdt.h>
#include <sys/time.h>
#include <unistd.h>
#include <stdio.h>

__extension__ unsigned short tracetest_testprobe_semaphore __attribute__ ((unused)) __attribute__ ((section (".probes"))) __attribute__ ((visibility ("hidden")));
__extension__ unsigned short tracetest_testprobe2_semaphore __attribute__ ((unused)) __attribute__ ((section (".probes"))) __attribute__ ((visibility ("hidden")));
__extension__ unsigned short tracetest2_testprobe2_semaphore __attribute__ ((unused)) __attribute__ ((section (".probes"))) __attribute__ ((visibility ("hidden")));

static long
myclock() {
  struct timeval tv;
  gettimeofday(&tv, NULL);
  printf("tracetest_testprobe_semaphore: %d\n", tracetest_testprobe_semaphore);
  printf("tracetest_testprobe2_semaphore: %d\n", tracetest_testprobe2_semaphore);
  printf("tracetest2_testprobe2_semaphore: %d\n", tracetest2_testprobe2_semaphore);
  DTRACE_PROBE2(tracetest,  testprobe,  tv.tv_sec, "Hello world");
  DTRACE_PROBE2(tracetest,  testprobe2, tv.tv_sec, "Hello world2");
  DTRACE_PROBE2(tracetest2, testprobe2, tv.tv_sec, "Hello world3");
  return tv.tv_sec;
}

int
main(int argc, char **argv) {
  while(1) {
    usleep(500000);
    myclock();
  }
    return 0;
}
```
You can see I added some `printf` statements and modified the code to run in an infinite loop.  When the probe is not enabled, you see output like this:

```
tracetest_testprobe_semaphore: 0
tracetest_testprobe2_semaphore: 0
tracetest2_testprobe2_semaphore: 0
```

If a probe (let's say `testprobe`) is enabled, you should see this:
```
tracetest_testprobe_semaphore: 1
```

In fact, if I use the `bcc` tool `trace`, that's exactly what happens:

```
$ sudo trace 'u:/home/cornelius/sources/bpftrace/build/tests/testprogs/usdt_test:testprobe' -p $(pidof usdt_test) -T

...
tracetest_testprobe_semaphore: 1
```
But when I try to do the same with `bpftrace`:

```
$ sudo src/bpftrace -e 'usdt::tracetest:testprobe { printf("%s\n", probe); }' -p `pidof usdt_test`
....
tracetest_testprobe_semaphore: 0
```

# Diagnosis

Using `strace` I traced the syscalls that `bpftrace` is making, and I found this bizarre sequence:

```
openat(AT_FDCWD, "/proc/24929/mem", O_RDWR) = 13                                                                    
...
write(13, "\1\0", 2)                    = 2
... (About half a dozen syscalls later)
openat(AT_FDCWD, "/proc/24929/mem", O_RDWR) = 13
...
write(13, "\0\0", 2)                    = 2
```

In other words, `bpftrace` is reaching into the `usdt_test` process memory, correctly setting the semaphore to `0x0100`, and then quickly thereafter setting it back to `0x0000` again.

After impugning the reputation of `bpftrace`'s mother most impolitely, I calmed down and decided to use `bpftrace` itself to gather more data.

Though I reproduced this problem with the versions packaged in the Arch linux repos, I then built both `bcc` and `bpftrace` from source using the latest `master`, in debug mode.  Then I used this (probably not very well written) eBPF program:

```ebpf
tracepoint:syscalls:sys_enter_openat {
 if (str(args->filename) == "/proc/17904/mem") {
     @is_interesting[tid] = 1
 }
}

tracepoint:syscalls:sys_exit_openat /@is_interesting[tid]/ {
 @fd[args->ret] = 1;
 delete(@is_interesting[tid]);
}

tracepoint:syscalls:sys_enter_write {
 if (@fd[args->fd]) {
     printf("Writer:\n%s\n", ustack());
 }
}

tracepoint:syscalls:sys_enter_close {
    delete(@fd[args->fd]);
}
```

It dumps the user stack trace of the calling process whenever it tries to write to the process memory of the `usdt_test` process.  This is what it produced on my system:

```

Writer:

        __write+24
        USDT::Probe::enable(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+108
        USDT::Context::enable_probe(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+431
        bcc_usdt_enable_fully_specified_probe+228
        bpftrace::AttachedProbe::attach_usdt(int)+794
        bpftrace::AttachedProbe::AttachedProbe(bpftrace::Probe&, std::tuple<unsigned char*, unsigned long>, int)+143
        std::_MakeUniq<bpftrace::AttachedProbe>::__single_object std::make_unique<bpftrace::AttachedProbe, bpftrace::Probe&, std::tuple<unsigned char*, unsigned long> const&, int&>(bpftrace::Probe&, std::tuple<unsigned char*, unsigned long> const&, int&)+137
        bpftrace::BPFtrace::attach_probe(bpftrace::Probe&, bpftrace::BpfOrc const&)+761
        bpftrace::BPFtrace::run(std::unique_ptr<bpftrace::BpfOrc, std::default_delete<bpftrace::BpfOrc> >)+794
        main+4312
        __libc_start_main+243
        0x5541f689495641d7

Writer:

        __write+24
        USDT::Probe::disable()+146
        USDT::Context::~Context()+193
        bcc_usdt_close+45
        bpftrace::AttachedProbe::attach_usdt(int)+1254
        bpftrace::AttachedProbe::AttachedProbe(bpftrace::Probe&, std::tuple<unsigned char*, unsigned long>, int)+143
        std::_MakeUniq<bpftrace::AttachedProbe>::__single_object std::make_unique<bpftrace::AttachedProbe, bpftrace::Probe&, std::tuple<unsigned char*, unsigned long> const&, int&>(bpftrace::Probe&, std::tuple<unsigned char*, unsigned long> const&, int&)+137
        bpftrace::BPFtrace::attach_probe(bpftrace::Probe&, bpftrace::BpfOrc const&)+761
        bpftrace::BPFtrace::run(std::unique_ptr<bpftrace::BpfOrc, std::default_delete<bpftrace::BpfOrc> >)+794
        main+4312
        __libc_start_main+243
        0x5541f689495641d7
```

Note this is the same pattern from the `strace` output: The first one is the correct on, enabling the probe, and the second one right after it is disabling it again.  Note the call to `bcc_usdt_close`.  That can't be right...

I did a bit of `git blame`ing and found [this commit from @dalehamel  about a month ago](https://github.com/iovisor/bpftrace/commit/ca814921f6db19b350b57d8e997507acf8cd0739#diff-ebadb41111e6bed6f2d85796dae909f1):

If I'm reading that right, this `bcc_usdt_close()` call was added.  When that call happens, the `bcc` library decrements the semaphore again, thus causing this defect.

# Fix

I assume this `bcc_usdt_close()` call was added for some reason.  I'm hoping @dalehamel  sees this and can comment.

Going forward I would propose that another testprog be added, something like the modification I made to `usdt_test.c`, to ensure USDT probes which use semaphores are exercised properly as part of the automated testing regime.

**Comment by dalehamel** [comment]
@anelson thanks for the detailed bug, I will try and take a look this weekend

**Comment by anelson** [comment]
Cool thanks @dalehamel.  FWIW I commented out this line in `attach_usdt`:

```c++
bcc_usdt_close(ctx);
```

and can confirm that fixed the problem.  Unfortunately it looks like that will leak some memory so it's not an idea long-term solution.  Maybe the `ctx` should be added to `AttachedProbe` and the `bcc_usdt_close` call made in the dtor...

**Comment by dalehamel** [comment]
Hi, sorry for the delay - I should be able to get to this during the upcoming week but have a few other things in the queue.

**Comment by anelson** [comment]
No worries Dale I'm sure you're busy.  Let me know if I can provide
anything else to make it easier for you.

**Comment by dalehamel** [comment]
@anelson finally got to this, sorry for the long delay.

Your great bug report makes this easier. I concur that the `bcc_usdt_close(ctx)` was mistakenly added.

I'll look for a more appropriate place to close it (the previous code doesn't appear to do so at all, so reverting it to start with), and I'll update the runtime tests to cover the semaphore case.

**Comment by dalehamel** [comment]
#691 up to fix this and introduce regression test

**Comment by anelson** [comment]
Outstanding @dalehamel, thanks for taking care of this!
